test/minikube/minikube_external_access_test.go

```diff
@@ -0,0 +1,234 @@
+// +build long
+
+package minikube
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"regexp"
+	"strconv"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/nuodb/nuodb-helm-charts/v3/test/testlib"
+
+	"github.com/gruntwork-io/terratest/modules/helm"
+	"github.com/gruntwork-io/terratest/modules/random"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+func findServiceNodePortE(service *corev1.Service, portName string) (int32, error) {
+	for _, port := range service.Spec.Ports {
+		if port.Name == portName {
+			return port.NodePort, nil
+		}
+	}
+	return 0, errors.New(
+		fmt.Sprintf("Unable to find NodePort for service %s", service.Name))
+}
+
+func findDomainExternalInfo(t *testing.T, namespaceName string) (string, int32) {
+	var domainAddress string
+	var domainPort int32
+	domainService := testlib.GetService(t, namespaceName, "nuodb-balancer")
+	if domainService.Spec.Type == corev1.ServiceTypeNodePort {
+		// get the internal address of the first Kubernetes node
+		for _, address := range testlib.GetNodesInternalAddresses(t) {
+			domainAddress = address
+			break
+		}
+		var err error
+		// get the nodePort of the domain service
+		domainPort, err = findServiceNodePortE(domainService, "48004-tcp")
+		require.NoError(t, err)
+	} else {
+		domainAddress = domainService.Status.LoadBalancer.Ingress[0].IP
+		domainPort = 48004
+	}
+	return domainAddress, domainPort
+}
+
+func verifyNuoSQLEngine(t *testing.T, namespaceName string, databaseName string,
+	properties map[string]string, expectedNodeIds []int32) {
+
+	address, port := findDomainExternalInfo(t, namespaceName)
+	pathToNuoSQL, ok := os.LookupEnv("NUOSQL_PATH")
+	if !ok {
+		pathToNuoSQL = "nuosql"
+	}
+	cmd := fmt.Sprintf("echo \"select GETNODEID() as NODEID from dual;\" | %s %s@%s:%d --user dba --password secret --vertical-output",
+		pathToNuoSQL, databaseName, address, port)
+	for key, value := range properties {
+		cmd += fmt.Sprintf(" --connection-property %s='%s'", key, value)
+	}
+	t.Logf("Running command: <%s>", cmd)
+	out, err := exec.Command("sh", "-c", cmd).Output()
+	if err != nil {
+		if exiterr, ok := err.(*exec.ExitError); ok {
+			// nuosql has exited with an non zero exit code; log command stderr
+			require.NoError(t, err, string(exiterr.Stderr))
+		}
+		require.NoError(t, err)
+	}
```

Comment:
I'm confused about this block. `err` is not `nil` but still not an error?

Reply:
If we get into the block then there is an error. I was just using `require.NoError()` to generate the usual message and fail the test. Is there a better way to do that? 
I see the `require.NoError()` internally calls `t.FailNow()` so an alternative would be something like:

```
t.Logf("nuosql command failed: %s", string(exiterr.Stderr))
t.FailNow()
```

Reply:
Okay, I was just wondering if the `err` check is ternary, because it could either be `nil` or some non-`nil` value that does not indicate an error, but that does not appear to be the case.

You are basically checking `err != nil` explicitly to guard the typecast, right? Maybe just add a comment to make it clear that this is structured like this to generate a better message.

Reply:
From my reading of this block:
* if there is an error (err != nil)
* then check to see if the err is a specific type of err (exec.ExitError)
* and if it _is_ an ExitError, then cause the stderr from that ExitErr to be output
* otherwise, just output the standard details from the err.

Reply:
> You are basically checking err != nil explicitly to guard the typecast, right

Yes, this was my idea, however, when I tested in a bit, it looks like that the cast doesn't need this guard. It will just fail if the `err` is `nil`.  You can see the small example that I used in https://play.golang.org/p/siBFV37YkCw.
I will reformat the code so that it's more readable and simple:

```golang
	out, err := exec.Command("sh", "-c", cmd).Output()
	if exiterr, ok := err.(*exec.ExitError); ok {
		// nuosql has exited with an non zero exit code; generate better error
		// message by including the command stderr
		require.NoError(t, err, string(exiterr.Stderr))
	} else {
		require.NoError(t, err)
	}
```